### PR TITLE
ci: testing with py3.11

### LINKED ISSUE #14652: Support Python 3.11
## 🚀 Feature
This is a tracking issue to support Python 3.11 released on 2022-10-03.

### Additional context
- [PEP 664 – Python 3.11 Release Schedule](https://peps.python.org/pep-0664/)
- [What’s New In Python 3.11](https://docs.python.org/3.11/whatsnew/3.11.html)

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/Lightning-AI/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/Lightning-AI/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/Lightning-AI/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/Lightning-AI/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging PyTorch Lightning, Transformers, and Hydra.

cc @borda @carmocca

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

Fixes #14652

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

cc @borda @carmocca @justusschock @awaelchli

### CODE REVIEW

**Comment by Borda** [comment]
some dependencies could be updated much sooner like `onnixruntime`, encreasing queue in #17823

**Review by awaelchli — APPROVED** [review]

**Comment by Borda** [comment]
could be related https://github.com/cython/cython/issues/4461

File: src/lightning/pytorch/trainer/connectors/data_connector.py [hunk 1]
```diff
@@ -400,14 +400,14 @@ def _check_dataloader_iterable(
         prefix = "train_" if trainer_fn == TrainerFn.FITTING else ""
         if not source.is_module():
             raise TypeError(
-                f"An invalid dataloader was passed to `Trainer.{trainer_fn}({prefix}dataloaders=...)`."
+                f"An invalid dataloader was passed to `Trainer.{trainer_fn.value}({prefix}dataloaders=...)`."
```
Comment (awaelchli) [inline src/lightning/pytorch/trainer/connectors/data_connector.py:403]: I made the enum changes because in Python 3.11, the string representation of the enum changed.

File: tests/tests_pytorch/utilities/test_exceptions.py [hunk 2]
```diff
@@ -14,54 +14,31 @@
...
-from lightning.pytorch.utilities.imports import _PYTHON_GREATER_EQUAL_3_11_0
 
 
 def test_augment_message():
     # exception without args
     exception = Exception()
     _augment_message(exception, "", "new message")
     assert not exception.args
-    if _PYTHON_GREATER_EQUAL_3_11_0:
-        assert not exception.__notes__
```
Comment (awaelchli) [inline tests/tests_pytorch/utilities/test_exceptions.py:26]: I forgot to adjust this test in #18399. I didn't notice because the CI wasn't running 3.11 yet.

Comment (Borda) [reply tests/tests_pytorch/utilities/test_exceptions.py:26]: you can do it here :innocent:

**Review by awaelchli — APPROVED** [review]

File: .github/workflows/ci-tests-fabric.yml [hunk 3]
```diff
@@ -47,10 +47,14 @@ jobs:
           - {os: "windows-2022", pkg-name: "lightning", python-version: "3.8", pytorch-version: "1.11"}
           - {os: "windows-2022", pkg-name: "lightning", python-version: "3.9", pytorch-version: "1.12"}
           - {os: "windows-2022", pkg-name: "lightning", python-version: "3.10", pytorch-version: "1.13"}
-          # only run PyTorch latest with Python latest
+          # only run PyTorch latest with Python recent
           - {os: "macOS-11", pkg-name: "lightning", python-version: "3.10", pytorch-version: "2.0"}
           - {os: "ubuntu-20.04", pkg-name: "lightning", python-version: "3.10", pytorch-version: "2.0"}
           - {os: "windows-2022", pkg-name: "lightning", python-version: "3.10", pytorch-version: "2.0"}
+          # only run PyTorch latest with Python latest, use Fabric scope to limit dependency issues
```
Comment (carmocca) [inline .github/workflows/ci-tests-fabric.yml:54]: These changes are missing in checkgroup

File: .github/workflows/ci-tests-pytorch.yml [hunk 4]
```diff
@@ -51,10 +51,14 @@ jobs:
           - {os: "windows-2022", pkg-name: "lightning", python-version: "3.8", pytorch-version: "1.11"}
           - {os: "windows-2022", pkg-name: "lightning", python-version: "3.9", pytorch-version: "1.12"}
           - {os: "windows-2022", pkg-name: "lightning", python-version: "3.10", pytorch-version: "1.13"}
-          # only run PyTorch latest with Python latest
+          # only run PyTorch latest with Python recent
           - {os: "macOS-11", pkg-name: "lightning", python-version: "3.10", pytorch-version: "2.0"}
           - {os: "ubuntu-20.04", pkg-name: "lightning", python-version: "3.10", pytorch-version: "2.0"}
           - {os: "windows-2022", pkg-name: "lightning", python-version: "3.10", pytorch-version: "2.0"}
+          # only run PyTorch latest with Python latest, use PyTorch scope to limit dependency issues
```
Comment (carmocca) [inline .github/workflows/ci-tests-pytorch.yml:58]: These changes are missing in checkgroup

File: .github/workflows/ci-tests-fabric.yml (hunk 3 above)
Comment (Borda) [reply .github/workflows/ci-tests-fabric.yml:54]: I was thinking of adding them later but fine to add now...

**Review by justusschock — APPROVED** [review]
